test/spmd/test_tracing.py

```diff
@@ -172,5 +178,134 @@ def test_all_gather_nd(self):
         self._test_all_gather_nd(torch.arange(4).reshape(2, 2))
 
 
+class TraceDistTensorTest(DistTensorTestBase):
+    @property
+    def world_size(self):
+        return 2
+
+    def _test_expand(self, xd, yd, f, mesh, out_spec):
+        # trace local graph
+        x = xd.redistribute(
+            device_mesh=mesh, placements=[Replicate()]
+        )._local_tensor
+        y = yd.redistribute(
+            device_mesh=mesh, placements=[Replicate()]
+        )._local_tensor
+        traced_f = make_fx(f)(x, y)
+
+        # map intermediate tensors in traced graph to DTensor objects
+        node_to_obj = {}
+
+        # map place holder to real input DTensor objects
+        def name_to_input(name):
+            return xd if "x" in name else yd
+
+        # avoid tracing _op_to_rules and _custom_dispatch_ops as placeholders
+        dispatch = partial(
+            operator_dispatch,
+            op_to_rules=DTensor._op_to_rules,
+            custom_dispatch_ops=DTensor._custom_dispatch_ops,
+        )
+
+        replacements = {}
+
+        # walk through the traced local graph and expand node with DTensor's
+        # dispatch implementation
+        for node in traced_f.graph.nodes:
+            if node.op == "placeholder":
+                node_to_obj[node] = name_to_input(node.name)
+            elif isinstance(node.target, torch._ops.OpOverload):
+                args = tree_map(lambda n: node_to_obj[n], node.args)
+                kwargs = tree_map(lambda n: node_to_obj[n], node.kwargs)
+
+                out = operator_dispatch(
```

Comment:
So this is doing "concrete" tracing, where we're tracing based on a concrete DTensor. Is this just for testing purposes or do we want to actually rely on concrete tracing for the final solution?

Reply:
This is not yet tracing. It is getting the intermediate DTensor outputs, so that we know what DTensor specs to pass to the next operator. Tracing is on line 318, which only applies to local tensors.

Reply:
Got it, yeah what I mean is that this corresponds to doing concrete tracing -- where you need the concrete tensors in order to propagate sharding.